### PR TITLE
Explore random forest models

### LINKED ISSUE #5: Explore Random Forest regression
We want to explore [Random Forest Regressor](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html) on hybrid labs data. For that, a python module needed to be called to fit, and predict. The module should:
- [ ] similar to [pycaret](https://pycaret.gitbook.io/docs) to compare different models, see [examples](https://nbviewer.org/github/pycaret/pycaret/blob/master/tutorials/Tutorial%20-%20Regression.ipynb)
- [ ] have a notebook in `docs` folder to show a simple example
- [ ] have tests
- [ ] have an easy interface and run fast (to be discussed)

**Data:**
The data folder `exp699_032024_TUDelft` is available on **Atlas** SharePoint at `Documents > HybridLabs > Example_data`. **NOTE:** The data folder `exp699_032024_TUDelft` is **not** public and cannot be shared with others. The trained model on this data cannot be also shared. For now, no need to store the trained model. 

The data folder `exp699_032024_TUDelft` includes:

- `readme.md`: it contains the experiment details and ML input/output
- `channel.csv:` it contains the variable names and units 
- `exp699.mat:` it contains the data and it is in matlab format.

For reading data in python, see [this notebook](https://github.com/hybridlabs-nl/FOWT-ML/blob/main/docs/notebooks/compare_linear_models.ipynb).

**Literature:**

Some literature is available at on **Atlas** SharePoint at `Documents > HybridLabs > Literature`.

### PR DESCRIPTION
**Pull request by fnattino**
Closes #5 

Concerning the performance of the predictions using random tree-based models (and others), I remembered a talk that mentioned tools to make model inference more efficient by creating "compiled" models in lower-level languages. I list the tools here, maybe they can become relevant later on, if the performance of the inference becomes crucial: 
* [TL2cgen](https://tl2cgen.readthedocs.io/): "model compiler for decision tree models." "You can convert any decision tree models (random forests, gradient boosting models) into C code and distribute it as a native binary."
* [hummingbird](https://github.com/microsoft/hummingbird) (by Microsoft): "a library for compiling trained traditional ML models into tensor computations." "Currently, you can use Hummingbird to convert your trained traditional ML models into [PyTorch](https://pytorch.org/), [TorchScript](https://pytorch.org/docs/stable/jit.html), [ONNX](https://onnx.ai/), and [TVM](https://docs.tvm.ai/)). Hummingbird [supports](https://github.com/microsoft/hummingbird/wiki/Supported-Operators) a variety of ML models and featurizers. These models include [scikit-learn](https://scikit-learn.org/stable/) Decision Trees and Random Forest, and also [LightGBM](https://github.com/Microsoft/LightGBM) and [XGBoost](https://github.com/dmlc/xgboost) Classifiers/Regressors. Support for other neural network backends and models is on our [roadmap](https://github.com/microsoft/hummingbird/wiki/Roadmap-for-Upcoming-Features-and-Support)."
* [lleaves](https://github.com/siboehm/lleaves): "lleaves converts trained LightGBM models to optimized machine code, speeding-up prediction by ≥10x."

### CODE REVIEW

**Comment by fnattino** [comment]
@SarahAlidoost I have added a module and the corresponding notebook for random forest - can you have a look and see whether what I did so far makes sense? If so, I will then add tests to wrap this up!

File: src/fowt_ml/ensemble.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+from typing import Protocol
+from typing import Self
+from numpy.typing import ArrayLike
+from sklearn.ensemble import ExtraTreesRegressor
+from sklearn.ensemble import RandomForestRegressor
+from sklearn.metrics import get_scorer
+from sklearn.model_selection import cross_val_score
+
+
+class Estimator(Protocol):
```
Comment (SarahAlidoost) [inline src/fowt_ml/ensemble.py:12]: I like the idea of defining the `Estimator` class. Instead of `Protocol`, let's use sklearn classes `RegressorMixin` and `BaseEstimator`, see the example [here](https://scikit-learn.org/1.6/modules/generated/sklearn.base.RegressorMixin.html).

File: src/fowt_ml/ensemble.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    def fit(self, X: ArrayLike, y: ArrayLike) -> Self: ...  # noqa: D102
+    def predict(self, X: ArrayLike) -> ArrayLike: ...  # noqa: D102
+
+
+ENSEMBLE_REGRESSORS = {
+    "ExtraTrees": ExtraTreesRegressor,
+    "RandomForest": RandomForestRegressor,
+}
```
Comment (SarahAlidoost) [inline src/fowt_ml/ensemble.py:20]: Can we move `ENSEMBLE_REGRESSORS` to the class `EnsembleModel`?, this way it would be the class property.

File: src/fowt_ml/ensemble.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class EnsembleModel:
+    """Class to handle random forest models and metrics for comparison."""
+
+    def __init__(self, estimator: str | Estimator, **kwargs: dict[str, Any]) -> None:
```
Comment (SarahAlidoost) [inline src/fowt_ml/ensemble.py:26]: nice to see that we can pass `kwargs` :+1:

File: src/fowt_ml/ensemble.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    def fit(self, X: ArrayLike, y: ArrayLike, sample_weight: ArrayLike = None) -> None:
+        """Fit an estimator."""
+        self.estimator.fit(X, y, sample_weight=sample_weight)
+        self._is_fitted = True
+
+    def predict(self, X: ArrayLike) -> ArrayLike:
+        """Predict with a fitted estimator."""
+        if not self._is_fitted:
+            raise RuntimeError("Run `model.fit(...)` first!")
+        return self.estimator.predict(X)
```
Comment (SarahAlidoost) [inline src/fowt_ml/ensemble.py:58]: In the beginning, I thought about this structure too, but I realized I was just wrapping sklearn functions in the new class without really adding anything. Since `self.estimator` is exposed, we can call its functions directly, so we don’t need to implement them. What do you think? :thinking:

File: src/fowt_ml/ensemble.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    def score(
```
Comment (SarahAlidoost) [inline src/fowt_ml/ensemble.py:60]: Same comment here, can we add `self.estimator.fit` to it? That way, we can get the score in one go. Also, I’d suggest renaming `score` to something like `calculate_score` to avoid matching sklearn’s API exactly, which could be confusing.

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@fnattino Thanks for implementing this module! Cool to see new ideas I hadn’t considered. I'd also like to add the `Estimator` class for linear models! 😄 I left some comments, let me know if anything isn’t clear. For the notebook, just one thing: could you please add `ExtraTrees` so we can compare the results? Thanks!

File: src/fowt_ml/ensemble.py (hunk 1 above)
Comment (fnattino) [reply src/fowt_ml/ensemble.py:12]: Uhm, as I understand sklearn's [user guide](https://scikit-learn.org/1.6/developers/develop.html#developing-scikit-learn-estimators) `BaseEstimator` and  `RegressorMixin` can be used to build new estimator objects - which I as far as I understand it's not what we want to do via FOWT-ML, right? 

What I wanted to do here is to have `EnsembleModel` to accept any valid estimator object - so something that implements the `fit()` and `predict()` methods. So this is why I have defined this "protocol" with methods and signatures. An alternative way was to have `BaseEstimator` as type for the input `estimator`, but I thought it would be good to leave the actual type open.

But maybe worth discussing it offline.

File: src/fowt_ml/ensemble.py (hunk 4 above)
Comment (fnattino) [reply src/fowt_ml/ensemble.py:58]: Yes, we share the same struggle 😩. Indeed, I am also not 100% sure on whether this is useful to copy sklearn API - maybe only because we can then log parameters/run sanity checks?

File: src/fowt_ml/ensemble.py (hunk 5 above)
Comment (fnattino) [reply src/fowt_ml/ensemble.py:60]: Good point, I will add a method to fit and score at once.

**Comment by fnattino** [comment]
This should now be done @SarahAlidoost  - I will merge main here after #15 is merged so that the fix to the sonar-cloud issues are solved!

File: src/fowt_ml/ensemble.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        X: ArrayLike,
+        y: ArrayLike,
+        cv: int | None = None,
+        scoring: str | None = None,
+    ) -> float | ArrayLike:
+        """Get Cross Validation score.
+
+        Scoring paramers overview: https://scikit-learn.org/stable/modules/model_evaluation.html#string-name-scorers
+        """  # noqa: E501
+        return cross_val_score(self.estimator, X, y, cv=cv, scoring=scoring)
```
Comment (SarahAlidoost) [inline src/fowt_ml/ensemble.py:45]:
```suggestion
```

Comment (SarahAlidoost) [reply src/fowt_ml/ensemble.py:45]: I think we decided not to wrap this function up too since we’re not adding anything to it.

**Review by SarahAlidoost — APPROVED** [review]
@fnattino thank you, it looks very good. just a minor comment. After removing the `cross_val_score`, we can merge the PR.

File: src/fowt_ml/ensemble.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        y_train: ArrayLike,
+        X_test: ArrayLike,
+        y_test: ArrayLike,
+        scoring: str | Iterable | None = None,
+    ) -> float | ArrayLike:
+        """Fit and calculate a score.
+
+        Scoring paramers overview: https://scikit-learn.org/stable/modules/model_evaluation.html#string-name-scorers
+        """  # noqa: E501
+        self.estimator.fit(X_train, y_train)
```
Comment (SarahAlidoost) [inline src/fowt_ml/ensemble.py:78]:
```suggestion
        self.model = self.estimator.fit(X_train, y_train)
```

Comment (SarahAlidoost) [reply src/fowt_ml/ensemble.py:78]: can we keep `self.model `, in case we want to save the fitted model later.
